File: src/test/java/application/browser/StackHelper.java
package application.browser;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

import static java.util.Collections.singletonList;
import static java.util.stream.Collectors.toList;

public class StackHelper extends AbstractCardContainerHelper {
  private static final Logger LOGGER = LoggerFactory.getLogger(StackHelper.class);

  StackHelper(MatagBrowser matagBrowser) {
    super(matagBrowser);
  }

  @Override
  protected WebElement containerElement() {
    return matagBrowser.findElement(By.id("stack"));
  }

  public void containsAbility(String ability) {
    containsAbilitiesExactly(singletonList(ability));
  }

  public void containsAbilitiesExactly(List<String> expectedTriggeredAbilities) {
    matagBrowser.wait(driver -> {
      var actualTriggeredAbilities = triggeredAbilities(containerElement());
      LOGGER.info("actualTriggeredAbilities={}   expectedTriggeredAbilities={}", actualTriggeredAbilities, expectedTriggeredAbilities);
      return expectedTriggeredAbilities.equals(actualTriggeredAbilities);
    });
  }

  private List<String> triggeredAbilities(WebElement containerElement) {
    return containerElement.findElements(By.className("triggered-ability")).stream()
      .map(WebElement::getText)
      .map(text -> text.replace("\n", " "))
      .collect(toList());
  }
}

File: src/test/kotlin/application/browser/StackHelper.kt
package application.browser

import org.openqa.selenium.By
import org.openqa.selenium.WebElement
import org.openqa.selenium.support.ui.ExpectedCondition
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class StackHelper internal constructor(matagBrowser: MatagBrowser) : AbstractCardContainerHelper(matagBrowser) {
    override fun containerElement(): WebElement {
        return matagBrowser.findElement(By.id("stack"))
    }

    override fun isEmpty() {
        super.isEmpty() // no cards on the stack
        toHaveAbilitiesSize(0) // no abilities on the stack
    }

    fun containsAbility(ability: String) {
        containsAbilitiesExactly(listOf(ability))
    }

    fun containsAbilitiesExactly(expectedTriggeredAbilities: List<String>) {
        matagBrowser.wait(ExpectedCondition {
            val actualTriggeredAbilities = triggeredAbilities(containerElement())
            LOGGER.info("actualTriggeredAbilities={}   expectedTriggeredAbilities={}", actualTriggeredAbilities, expectedTriggeredAbilities)
            expectedTriggeredAbilities == actualTriggeredAbilities
        })
    }

    fun toHaveAbilitiesSize(size: Int) {
        matagBrowser.wait(ExpectedCondition {
            val actualTriggeredAbilities = triggeredAbilities(containerElement())
            if (actualTriggeredAbilities.size == size) {
                return@ExpectedCondition true
            } else {
                LOGGER.info("Expected {} no abilities on the stack but got {} ({})", size, actualTriggeredAbilities.size, actualTriggeredAbilities)
                return@ExpectedCondition false
            }
        })
    }

    private fun triggeredAbilities(containerElement: WebElement): List<String> {
        return containerElement.findElements(By.className("triggered-ability"))
            .map { it.text }
            .map { it.replace("\n", " ") }
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(StackHelper::class.java)
    }
}
